src/Tests/Microsoft.Diagnostics.Monitoring.HostingStartup.UnitTests/ParameterCapturing/Pipeline/ParameterCapturingPipelineTests.cs

```diff
@@ -0,0 +1,292 @@
+        private readonly Action<Guid> _onCapturingStop;
+        private readonly Action<Guid, ParameterCapturingEvents.CapturingFailedReason, string> _onCapturingFailed;
+
+        public TestParameterCapturingCallbacks(
+            Action<StartCapturingParametersPayload, IList<MethodInfo>> onCapturingStart = null,
+            Action<Guid> onCapturingStop = null,
+            Action<Guid, ParameterCapturingEvents.CapturingFailedReason, string> onCapturingFailed = null)
+        {
+            _onCapturingStart = onCapturingStart;
+            _onCapturingStop = onCapturingStop;
+            _onCapturingFailed = onCapturingFailed;
+        }
+
+        public void CapturingStart(StartCapturingParametersPayload request, IList<MethodInfo> methods)
+        {
+            _onCapturingStart?.Invoke(request, methods);
+        }
+
+        public void CapturingStop(Guid RequestId)
+        {
+            _onCapturingStop?.Invoke(RequestId);
+        }
+
+        public void FailedToCapture(Guid RequestId, ParameterCapturingEvents.CapturingFailedReason Reason, string Details)
+        {
+            _onCapturingFailed?.Invoke(RequestId, Reason, Details);
+        }
+    }
+
+    [TargetFrameworkMonikerTrait(TargetFrameworkMonikerExtensions.CurrentTargetFrameworkMoniker)]
+    public class ParameterCapturingPipelineTests
+    {
+        private readonly ITestOutputHelper _outputHelper;
+
+        public ParameterCapturingPipelineTests(ITestOutputHelper outputHelper)
+        {
+            _outputHelper = outputHelper;
+        }
+
+        [Fact]
+        public void RequestStop_InvalidRequestId_Throws()
+        {
+            // Arrange
+            ParameterCapturingPipeline pipeline = new(new TestFunctionProbesManager(), new TestParameterCapturingCallbacks());
+
+            // Act & Assert
+            Assert.Throws<ArgumentException>(() => pipeline.RequestStop(Guid.NewGuid()));
+        }
+
+        [Fact]
+        public async Task Request_DoesInstallAndNotify()
+        {
+            // Arrange
+            TaskCompletionSource probeManagerStartSource = new(TaskCreationOptions.RunContinuationsAsynchronously);
+            TestFunctionProbesManager probeManager = new(
+                onStart: (_) =>
+                {
+                    probeManagerStartSource.TrySetResult();
+                });
+
+            TaskCompletionSource<Guid> onStartCallbackSource = new(TaskCreationOptions.RunContinuationsAsynchronously);
+            TestParameterCapturingCallbacks callbacks = new(
+                onCapturingStart: (payload, _) =>
+                {
+                    onStartCallbackSource.TrySetResult(payload.RequestId);
+                });
+
+            ParameterCapturingPipeline pipeline = new(probeManager, callbacks);
+            StartCapturingParametersPayload payload = CreateStartCapturingPayload(Timeout.InfiniteTimeSpan);
+
+            using CancellationTokenSource cts = new();
```

Comment:
For each of these tests, consider moving this to the beginning, setting a generous timeout on it, and cancelling each of the `TaskCompletionSource` instances when this expires to prevent tests from hanging forever.